File: src/weaponsprocurement/gui/WimGuiModalInput.java
package weaponsprocurement.gui;

import com.fs.starfarer.api.input.InputEventAPI;
import com.fs.starfarer.api.ui.CustomPanelAPI;
import org.lwjgl.input.Keyboard;

import java.util.List;

final class WimGuiModalInput<A> {
    private final WimGuiButtonPoller<A> buttonPoller;
    private CustomPanelAPI root;
    private WimGuiListBounds listBounds;

    WimGuiModalInput(int buttonPollFramesAfterMouseEvent) {
        this.buttonPoller = new WimGuiButtonPoller<A>(buttonPollFramesAfterMouseEvent);
    }

    void setRoot(CustomPanelAPI root) {
        this.root = root;
    }

    void setListBounds(WimGuiListBounds listBounds) {
        this.listBounds = listBounds;
    }

    List<WimGuiButtonBinding<A>> buttonBindings() {
        return buttonPoller.bindings();
    }

    void clearButtonBindings() {
        buttonPoller.clearBindings();
    }

    void clearCheckedButtons() {
        buttonPoller.clearCheckedButtons();
    }

    void suppressNextMouseUpPoll() {
        buttonPoller.suppressNextMouseUp();
    }

    WimGuiInputResult processInput(List<InputEventAPI> events) {
        if (events == null) {
            return WimGuiInputResult.none();
        }
        for (int i = 0; i < events.size(); i++) {
            InputEventAPI event = events.get(i);
            if (event == null || event.isConsumed()) {
                continue;
            }
            if (event.isKeyDownEvent() && event.getEventValue() == Keyboard.KEY_ESCAPE) {
                event.consume();
                return WimGuiInputResult.closeRequested();
            }
            if (event.isMouseScrollEvent() && event.getEventValue() != 0 && isMouseInList(event)) {
                int delta = WimGuiScroll.wheelDelta(event.getEventValue());
                event.consume();
                return WimGuiInputResult.scroll(delta);
            }
        }
        buttonPoller.requestPollFromInput(events);
        return WimGuiInputResult.none();
    }

    boolean processButtonsIfRequested(WimGuiButtonPoller.ActionHandler<A> handler) {
        return buttonPoller.processIfRequested(handler);
    }

    private boolean isMouseInList(InputEventAPI event) {
        return listBounds != null && listBounds.contains(root, event);
    }
}

File: src/main/kotlin/weaponsprocurement/gui/WimGuiModalInput.kt
package weaponsprocurement.gui

import com.fs.starfarer.api.input.InputEventAPI
import com.fs.starfarer.api.ui.CustomPanelAPI
import org.lwjgl.input.Keyboard

class WimGuiModalInput<A>(buttonPollFramesAfterMouseEvent: Int) {
    private val buttonPoller = WimGuiButtonPoller<A>(buttonPollFramesAfterMouseEvent)
    private var root: CustomPanelAPI? = null
    private var listBounds: WimGuiListBounds? = null

    fun setRoot(root: CustomPanelAPI?) {
        this.root = root
    }

    fun setListBounds(listBounds: WimGuiListBounds?) {
        this.listBounds = listBounds
    }

    fun buttonBindings(): MutableList<WimGuiButtonBinding<A>> = buttonPoller.bindings()

    fun clearButtonBindings() {
        buttonPoller.clearBindings()
    }

    fun clearCheckedButtons() {
        buttonPoller.clearCheckedButtons()
    }

    fun suppressNextMouseUpPoll() {
        buttonPoller.suppressNextMouseUp()
    }

    fun processInput(events: List<InputEventAPI>?): WimGuiInputResult {
        if (events == null) {
            return WimGuiInputResult.none()
        }
        for (event in events) {
            if (event.isConsumed) {
                continue
            }
            if (event.isKeyDownEvent && event.eventValue == Keyboard.KEY_ESCAPE) {
                event.consume()
                return WimGuiInputResult.closeRequested()
            }
            if (event.isMouseScrollEvent && event.eventValue != 0 && isMouseInList(event)) {
                val delta = WimGuiScroll.wheelDelta(event.eventValue)
                event.consume()
                return WimGuiInputResult.scroll(delta)
            }
        }
        buttonPoller.requestPollFromInput(events)
        return WimGuiInputResult.none()
    }

    fun processButtonsIfRequested(handler: WimGuiButtonPoller.ActionHandler<A>?): Boolean =
        buttonPoller.processIfRequested(handler)

    private fun isMouseInList(event: InputEventAPI): Boolean = listBounds?.contains(root, event) == true
}
